emacs/.emacs.d/settings.org
* Package management
#+BEGIN_SRC emacs-lisp
;; Set up ELPA and MELPA
(require 'package)
(setq package-archives '(("gnu"   . "https://elpa.gnu.org/packages/")
                         ("melpa" . "https://melpa.org/packages/")))
(package-initialize)

;; Set up package-use
(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))
(require 'use-package)
#+END_SRC
* General configuration
#+BEGIN_SRC emacs-lisp
(setq-default indent-tabs-mode nil)
(setq-default tab-width 3)
(setq indent-line-finction 'insert-tab)
(setq-default fill-column 67)
(global-auto-revert-mode t)
(setq backup-directory-alist `(("." . "~/.saves")))
#+END_SRC
* Org
#+BEGIN_SRC emacs-lisp
(require 'org)

(define-key global-map "\C-cl" 'org-store-link)
(define-key global-map "\C-cc" 'org-capture)

(setq org-log-done t)


(setq org-todo-keywords
  '((sequence
       "⚑(p)"
       "☐(t)"
       "|"
       "☑(d)"
       "☒(c)")))

(define-abbrev org-mode-abbrev-table "PROJECT"   "⚑")
(define-abbrev org-mode-abbrev-table "TODO"      "☐")
(define-abbrev org-mode-abbrev-table "DONE"      "☑")
(define-abbrev org-mode-abbrev-table "CANCELLED" "☒")
(add-hook 'org-mode-hook 'abbrev-mode)

(setq org-return-follows-link t)

(setq org-startup-folded 'content)
(setq org-indent-indentation-per-level 4)

(add-hook 'org-mode-hook (lambda ()
  (org-indent-mode 1)))


(setq org-todo-keyword-faces
  '(("⚑" . (:foreground "#af005f"
            :weight bold))
    ("☐" . (:foreground "#af8700"
            :weight bold))
    ("☑" . (:foreground "#5f8700"
            :weight bold))
    ("☒" . (:foreground "#af0000"
            :weight bold))))

(setq org-fontify-done-headline t)
(custom-set-faces
;; '(org-done ((t (:foreground "PaleGreen"   
;;                 :weight normal))))
 '(org-headline-done 
            ((((class color)
               (min-colors 16)
               (background dark)) 
               (:foreground "#585858" :strike-through t))))
)


; Remove leading asterisks & space from all headlines
(font-lock-add-keywords 'org-mode
  '(("^\\*+ "
      (0 (compose-region
           (match-beginning 0)
           (match-end 0)
           "")
         (put-text-property
           (match-beginning 0)
           (match-end 0)
           '(invisible t))))))
#+END_SRC
** Agenda
#+BEGIN_SRC emacs-lisp
(define-key global-map (kbd "C-c a") 'org-agenda)

(setq org-agenda-files '("~/p/gtd"))
(setq org-agenda-dim-blocked-tasks nil)
(setq org-agenda-compact-blocks t)
#+END_SRC
** Clock management
#+BEGIN_SRC emacs-lisp
; TODO: investigate implementation that uses org-tree-to-indirect-buffer

(setq focus-buffer "*Task*")
(setq focus-element nil)

(defun goto-task ()
  (when focus-element
  (goto-char (point-min))
  (search-forward
   (org-element-property :raw-value focus-element))))

(defun get-task-bounds ()
  (cons
   (org-element-property :begin focus-element)
   (- (org-element-property :end focus-element) 1)))

(defun task-focus ()
  (when (get-buffer focus-buffer)
    (kill-buffer focus-buffer))
  (setq focus-element (org-element-at-point (car org-clock-history)))
  (goto-task)
  (if (org-folded-p)
      (org-cycle))
  (let ((bounds (get-task-bounds)))
    (switch-to-buffer
      (clone-indirect-buffer focus-buffer nil))
    (narrow-to-region (car bounds) (cdr bounds)))
    (when (get-buffer "gtd")
          (kill-buffer "gtd")))

(defun org-folded-p ()
  (and (or (org-at-heading-p)
           (org-at-item-p))
       (invisible-p (point-at-eol))))

(defun task-unfocus ()
  (when (get-buffer focus-buffer)
    (kill-buffer focus-buffer)
    (goto-task)
    (if (not (org-folded-p))
        (org-cycle))))

(add-hook 'org-clock-in-hook 'task-focus)
(add-hook 'org-clock-out-hook 'task-unfocus)
#+END_SRC
** Journal
#+BEGIN_SRC emacs-lisp
(use-package org-journal
  :ensure t
  :after org
  :custom
  (org-journal-dir "~/p/journal")
  (org-journal-file-format "%Y-%m-%d.org")
  (org-journal-date-format "%Y-%m-%d")
  (org-journal-find-file 'find-file)
  (org-journal-enable-agenda-integration t))

(add-hook 'org-journal-mode-hook (lambda () (auto-fill-mode 1)))
#+END_SRC
** Babel
#+BEGIN_SRC emacs-lisp
(setq org-confirm-babel-evaluate nil)
#+END_SRC
*** Scheme
#+BEGIN_SRC emacs-lisp

#+END_SRC
*** Lilypond
#+BEGIN_SRC emacs-lisp
(org-babel-do-load-languages
 'org-babel-load-languages
 '((emacs-lisp . t)
   (org . t)
   (C . t)
   (plantuml . t)
   (latex . t)
   (lilypond . t)
   (scheme . t)))

(setq org-plantuml-jar-path (expand-file-name "/usr/share/java/plantuml/plantuml.jar"))
(add-to-list 'org-src-lang-modes '("plantuml" . plantuml))

;; Instruct org to prepend each lilypond source block
;; with some code that will remove the attribution at
;; the bottom of the exported result
;; 
;; Doesn't currently work, because the internals of
;; ob-lilypond screw with this at runtime
(add-to-list 'org-babel-default-header-args:lilypond
  '(:prologue . "\\header {\n tagline = \"\"\n}"))
#+END_SRC
** Export
#+BEGIN_SRC emacs-lisp
(require 'ox-latex)
(add-to-list 'org-latex-packages-alist '("" "minted"))
(setq org-latex-listings 'minted)

(setq org-latex-pdf-process
      '("xelatex -shell-escape -interaction nonstopmode -output-directory %o %f"
        "xelatex -shell-escape -interaction nonstopmode -output-directory %o %f"))

(setq org-src-fontify-natively t)

(setq org-latex-minted-options
      '(("bgcolor" "black")
        ("linenos" "true")
	    ("breaklines" "true")
	    ("breakanywhere" "true")
        ("frame" "lines")))
#+END_SRC

* EVIL
#+BEGIN_SRC emacs-lisp
(setq evil-want-C-i-jump nil)
(use-package evil
  :ensure t
  :defer .1
  :init
  (setq evil-search-module 'evil-search)
  :config
  ;; Remap hjkl -> htns
  (define-key evil-motion-state-map "t" 'evil-next-visual-line)
  (define-key evil-motion-state-map "n" 'evil-previous-visual-line)
  (define-key evil-motion-state-map "s" 'evil-forward-char)
  (unbind-key "s" evil-normal-state-map)

  ;; Move t -> j, n -> l, N -> L, s -> k
  (define-key evil-motion-state-map "j" 'evil-find-char-to)
  (define-key evil-motion-state-map "l" 'evil-ex-search-next)
  (define-key evil-motion-state-map "L" 'evil-ex-search-previous)
  (define-key evil-motion-state-map "k" 'evil-substitute)

  ;; Get C-c to escape to normal mode
  (define-key evil-insert-state-map (kbd "C-c") 'evil-normal-state)
  (evil-mode))
#+END_SRC
* EXWM
#+BEGIN_SRC emacs-lisp
(defun exwm-config ()
  (setq exwm-workspace-number 10)
  
  ;; Global keybindings
  (unless (get 'exwm-input-global-keys 'saved-value)
    (setq exwm-input-global-keys
          `(
            ;; 's-r': Reset (to line-mode).
            ([?\s-r] . exwm-reset)
            ;; 's-&': Launch application.
            ([?\s-d] . (lambda (command)
                         (interactive (list (read-shell-command "$ ")))
                         (start-process-shell-command command nil command)))
            ;; 's-N': Switch to certain workspace.
            ,@(mapcar (lambda (i)
                        `(,(kbd (format "s-%d" (mod (+ i 1) 10))) .
                          (lambda ()
                            (interactive)
                            (exwm-workspace-switch-create ,i))))
                      (number-sequence 0 9))
  
            ;; 's-S-N': Switch to certain workspace.
            ,@(mapcar (lambda (x)
                        `(,(kbd (format "s-%s" (cdr x))) .
                          (lambda ()
                            (interactive)
                            (exwm-workspace-move-window (- ,(car x) 1)))))
					  '((1  . "!")
					    (2  . "@")
						(3  . "#")
						(4  . "$")
						(5  . "%")
						(6  . "^")
						(7  . "&")
						(8  . "*")
						(9  . "(")
						(10 . ")"))))))
  
  (require 'exwm-systemtray)
  (exwm-systemtray-enable))

;;(use-package exwm
;;  :ensure t
;;  :config
;;  (exwm-config)
;;  (exwm-enable))
#+end_src
* Visual customization
** disable menubar, toolbar, and scrollbar
#+begin_src emacs-lisp
(menu-bar-mode -1)
(tool-bar-mode -1)
(scroll-bar-mode -1)
#+end_src
** Theme
#+BEGIN_SRC emacs-lisp
(use-package solarized-theme
  :defer t
  :ensure t
  :init
  (load-theme 'solarized-dark t))
#+END_SRC
** Font
#+BEGIN_SRC emacs-lisp
(add-to-list 'default-frame-alist '(font . "Inconsolata-14"))
#+END_SRC
** Line numbers
#+BEGIN_SRC emacs-lisp
(when (version<= "26.0.50" emacs-version)
  (global-display-line-numbers-mode))
#+END_SRC
* Window management
#+BEGIN_SRC emacs-lisp
; Moving focus between windows
;(use-package windmove
;  :ensure t
;  :bind
;  (("" . windmove-left)
;   ("" . windmove-down)
;   ("" . windmove-up)
;   ("" . windmove-right)))
#+END_SRC
* Major mode configuration
** C Mode
#+BEGIN_SRC emacs-lisp
(setq-default c-basic-offset 4
              tab-width 4
              indent-tabs-mode t
			  c-default-style "bsd")
#+END_SRC
** LaTex Mode
#+BEGIN_SRC emacs-lisp
(add-hook 'latex-mode-hook 'auto-fill-mode)
(add-hook 'latex-mode-hook (lambda () (electric-indent-mode -1)))
#+END_SRC
** EVIL Mode
#+BEGIN_SRC emacs-lisp
(setq evil-auto-indent nil)
#+END_SRC
** Markdown Mode
#+BEGIN_SRC emacs-lisp
(use-package markdown-mode
  :ensure t
  :mode (("README\\.md\\'" . gfm-mode)
         ("\\.md\\'" . markdown-mode))
  :init (setq markdown-command "pandoc"))
#+END_SRC
** Lilypond mode
#+BEGIN_SRC emacs-lisp
(setq load-path (append (list (expand-file-name "/usr/share/emacs/site-lisp")) load-path))
(autoload 'LilyPond-mode "lilypond-mode" "LilyPond Editing Mode" t)
(add-to-list 'auto-mode-alist '("\\.ly$" . LilyPond-mode))
(add-to-list 'auto-mode-alist '("\\.ily$" . LilyPond-mode))
(add-hook 'LilyPond-mode-hook (lambda () (turn-on-font-lock)))
#+END_SRC
** Paredit
#+BEGIN_SRC emacs-lisp
(use-package paredit
   :ensure t
   :commands (enable-paredit-mode)
   :init
   (add-hook 'emacs-lisp-mode-hook #'enable-paredit-mode))
#+END_SRC
** Company
#+BEGIN_SRC emacs-lisp
(use-package company
   :ensure t
   :bind (:map company-active-map
          ("C-n" . company-select-next)
          ("C-p" . company-select-previous))
   :config
   (setq company-idle-delay 0.3)
   (global-company-mode t))
#+END_SRC
** Geiser
#+BEGIN_SRC emacs-lisp
(use-package geiser
   :ensure t
   :defer t
   :init (add-hook 'scheme-mode-hook 'geiser-mode)
   :commands geiser-mode
   :config (setq geiser-active-implementations '(mit)))
#+END_SRC
** VHDL
#+BEGIN_SRC emacs-lisp
(setq vhdl-basic-offset 4)
#+END_SRC


emacs/.emacs.d/settings.el
;; Set up ELPA and MELPA
(require 'package)
(setq package-archives '(("gnu"   . "https://elpa.gnu.org/packages/")
                         ("melpa" . "https://melpa.org/packages/")))
(package-initialize)

;; Set up package-use
(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))
(require 'use-package)

(setq-default indent-tabs-mode nil)
(setq-default tab-width 3)
(setq indent-line-finction 'insert-tab)
(setq-default fill-column 67)
(global-auto-revert-mode t)
(setq backup-directory-alist `(("." . "~/.saves")))

(require 'org)

(define-key global-map "\C-cl" 'org-store-link)
(define-key global-map "\C-cc" 'org-capture)

(setq org-log-done t)


(setq org-todo-keywords
  '((sequence
       "⚑(p)"
       "☐(t)"
       "|"
       "☑(d)"
       "☒(c)")))

(define-abbrev org-mode-abbrev-table "PROJECT"   "⚑")
(define-abbrev org-mode-abbrev-table "TODO"      "☐")
(define-abbrev org-mode-abbrev-table "DONE"      "☑")
(define-abbrev org-mode-abbrev-table "CANCELLED" "☒")
(add-hook 'org-mode-hook 'abbrev-mode)

(setq org-return-follows-link t)

(setq org-startup-folded 'content)
(setq org-indent-indentation-per-level 4)

(add-hook 'org-mode-hook (lambda ()
  (org-indent-mode 1)))


(setq org-todo-keyword-faces
  '(("⚑" . (:foreground "#af005f"
            :weight bold))
    ("☐" . (:foreground "#af8700"
            :weight bold))
    ("☑" . (:foreground "#5f8700"
            :weight bold))
    ("☒" . (:foreground "#af0000"
            :weight bold))))

(setq org-fontify-done-headline t)
(custom-set-faces
;; '(org-done ((t (:foreground "PaleGreen"   
;;                 :weight normal))))
 '(org-headline-done 
            ((((class color)
               (min-colors 16)
               (background dark)) 
               (:foreground "#585858" :strike-through t))))
)


; Remove leading asterisks & space from all headlines
(font-lock-add-keywords 'org-mode
  '(("^\\*+ "
      (0 (compose-region
           (match-beginning 0)
           (match-end 0)
           "")
         (put-text-property
           (match-beginning 0)
           (match-end 0)
           '(invisible t))))))

(define-key global-map (kbd "C-c a") 'org-agenda)

(setq org-agenda-files '("~/p/gtd"))
(setq org-agenda-dim-blocked-tasks nil)
(setq org-agenda-compact-blocks t)

(use-package org-journal
  :ensure t
  :after org
  :custom
  (org-journal-dir "~/p/journal")
  (org-journal-file-format "%Y-%m-%d.org")
  (org-journal-date-format "%Y-%m-%d")
  (org-journal-find-file 'find-file)
  (org-journal-enable-agenda-integration t))

(add-hook 'org-journal-mode-hook (lambda () (auto-fill-mode 1)))

(setq org-confirm-babel-evaluate nil)



(org-babel-do-load-languages
 'org-babel-load-languages
 '((emacs-lisp . t)
   (org . t)
   (C . t)
   (plantuml . t)
   (latex . t)
   (lilypond . t)
   (scheme . t)))

(setq org-plantuml-jar-path (expand-file-name "/usr/share/java/plantuml/plantuml.jar"))
(add-to-list 'org-src-lang-modes '("plantuml" . plantuml))

;; Instruct org to prepend each lilypond source block
;; with some code that will remove the attribution at
;; the bottom of the exported result
;; 
;; Doesn't currently work, because the internals of
;; ob-lilypond screw with this at runtime
(add-to-list 'org-babel-default-header-args:lilypond
  '(:prologue . "\\header {\n tagline = \"\"\n}"))

(require 'ox-latex)
(add-to-list 'org-latex-packages-alist '("" "minted"))
(setq org-latex-listings 'minted)

(setq org-latex-pdf-process
      '("xelatex -shell-escape -interaction nonstopmode -output-directory %o %f"
        "xelatex -shell-escape -interaction nonstopmode -output-directory %o %f"))

(setq org-src-fontify-natively t)

(setq org-latex-minted-options
      '(("bgcolor" "black")
        ("linenos" "true")
	    ("breaklines" "true")
	    ("breakanywhere" "true")
        ("frame" "lines")))

(setq evil-want-C-i-jump nil)
(use-package evil
  :ensure t
  :defer .1
  :init
  (setq evil-search-module 'evil-search)
  :config
  ;; Remap hjkl -> htns
  (define-key evil-motion-state-map "t" 'evil-next-visual-line)
  (define-key evil-motion-state-map "n" 'evil-previous-visual-line)
  (define-key evil-motion-state-map "s" 'evil-forward-char)
  (unbind-key "s" evil-normal-state-map)

  ;; Move t -> j, n -> l, N -> L, s -> k
  (define-key evil-motion-state-map "j" 'evil-find-char-to)
  (define-key evil-motion-state-map "l" 'evil-ex-search-next)
  (define-key evil-motion-state-map "L" 'evil-ex-search-previous)
  (define-key evil-motion-state-map "k" 'evil-substitute)

  ;; Get C-c to escape to normal mode
  (define-key evil-insert-state-map (kbd "C-c") 'evil-normal-state)
  (evil-mode))

(defun exwm-config ()
  (setq exwm-workspace-number 10)
  
  ;; Global keybindings
  (unless (get 'exwm-input-global-keys 'saved-value)
    (setq exwm-input-global-keys
          `(
            ;; 's-r': Reset (to line-mode).
            ([?\s-r] . exwm-reset)
            ;; 's-&': Launch application.
            ([?\s-d] . (lambda (command)
                         (interactive (list (read-shell-command "$ ")))
                         (start-process-shell-command command nil command)))
            ;; 's-N': Switch to certain workspace.
            ,@(mapcar (lambda (i)
                        `(,(kbd (format "s-%d" (mod (+ i 1) 10))) .
                          (lambda ()
                            (interactive)
                            (exwm-workspace-switch-create ,i))))
                      (number-sequence 0 9))
  
            ;; 's-S-N': Switch to certain workspace.
            ,@(mapcar (lambda (x)
                        `(,(kbd (format "s-%s" (cdr x))) .
                          (lambda ()
                            (interactive)
                            (exwm-workspace-move-window (- ,(car x) 1)))))
					  '((1  . "!")
					    (2  . "@")
						(3  . "#")
						(4  . "$")
						(5  . "%")
						(6  . "^")
						(7  . "&")
						(8  . "*")
						(9  . "(")
						(10 . ")"))))))
  
  (require 'exwm-systemtray)
  (exwm-systemtray-enable))

;;(use-package exwm
;;  :ensure t
;;  :config
;;  (exwm-config)
;;  (exwm-enable))

(menu-bar-mode -1)
(tool-bar-mode -1)
(scroll-bar-mode -1)

(use-package solarized-theme
  :defer t
  :ensure t
  :init
  (load-theme 'solarized-dark t))

(add-to-list 'default-frame-alist '(font . "Droid Sans Mono-14"))

(when (version<= "26.0.50" emacs-version)
  (global-display-line-numbers-mode))

; Moving focus between windows
;(use-package windmove
;  :ensure t
;  :bind
;  (("" . windmove-left)
;   ("" . windmove-down)
;   ("" . windmove-up)
;   ("" . windmove-right)))

(setq-default c-basic-offset 4
              tab-width 4
              indent-tabs-mode t
			  c-default-style "bsd")

(add-hook 'latex-mode-hook 'auto-fill-mode)
(add-hook 'latex-mode-hook (lambda () (electric-indent-mode -1)))

(setq evil-auto-indent nil)

(use-package markdown-mode
  :ensure t
  :mode (("README\\.md\\'" . gfm-mode)
         ("\\.md\\'" . markdown-mode))
  :init (setq markdown-command "pandoc"))

(setq load-path (append (list (expand-file-name "/usr/share/emacs/site-lisp")) load-path))
(autoload 'LilyPond-mode "lilypond-mode" "LilyPond Editing Mode" t)
(add-to-list 'auto-mode-alist '("\\.ly$" . LilyPond-mode))
(add-to-list 'auto-mode-alist '("\\.ily$" . LilyPond-mode))
(add-hook 'LilyPond-mode-hook (lambda () (turn-on-font-lock)))

(use-package paredit
   :ensure t
   :commands (enable-paredit-mode)
   :init
   (add-hook 'emacs-lisp-mode-hook #'enable-paredit-mode))

(use-package company
   :ensure t
   :bind (:map company-active-map
          ("C-n" . company-select-next)
          ("C-p" . company-select-previous))
   :config
   (setq company-idle-delay 0.3)
   (global-company-mode t))

(use-package geiser
   :ensure t
   :defer t
   :init (add-hook 'scheme-mode-hook 'geiser-mode)
   :commands geiser-mode
   :config (setq geiser-active-implementations '(mit)))

(setq vhdl-basic-offset 4)
